<div align="center"><h1><b>Resolución de la Ecuación Diferencial de Calor</b></h1></div>
<div align="center"><h1><b>Utilizando el Lenguaje C y Paradigmas de Paralelización</b></h1></div>
<div align="center"><h3>Nicolás Cuello Suárez - Curso de Programación de Alto Desempeño</h3></div>
<div align="center"><h3>Facultad de Ciencias Básicas - Universidad de Medellín</h3></div>

---

## **Una Vista Sobre la Dinámica Térmica.**

Realicemos un ejercicio simple: Imagina una **barra de metal** cuya longitud es, digamos, del tamaño de una *regla escolar*. Con una mano, la sostienes por un **extremo**, con la otra, aproximas un encendedor otro extremo para calentarlo. No lo apoyas, el fuego no toca la barrra, solo lo acercas; al poco tiempo ves que el metal se calienta **justo donde está la llama**. Teniendo en cuenta esta situación, una pregunta válida que puede surgir es:

- **Si me quedo en esta posición, ¿cuánto tardará en sentirse el calor en la mano que uso para sostener el otro extremo?**

Interpolando lo anterior a cualquier otra situación que se relacione con el **fuego**, por simple **empiria**, sabemos que la propagación del fuego no es **instantánea** ni mucho menos, el calor viaja **de forma lenta** pero constante, no salta ni vuela, más bien **se difunde**. Este es un conocimiento básico y, como lo dije antes, empírico, pero, si somos curiosos y nos vamos más allá de lo que observamos, preguntas que pueden surgir son:

- **¿Podemos estimar, de alguna manera, el tiempo en el que el extremo frío se calentará?**
- **¿Existe alguna manera de conocer la temperatura de un punto cualquiera de la barra?**
- **¿Qué herramientas podemos utilizar en este caso?**

Este tipo de **curiosidades**, que en primera instancia uno pensaría que son tontas e inútiles, son las cuestiones que han motivado el avance **científico**, son preguntas completamente habituales y casi de obligada ejecución en un **contexto de ciencia**; son tan habituales que alguien, hace mucho tiempo atrás, en el siglo **XIX**, logró hallar una respuesta concreta y demostrable, el nombre es **Jean Fourier**, un matemático francés cuya mente fue una de las más brillantes en su época.

**Fourier** no solo encontró una resolución a estas dudas instrumentalizando un lenguaje universal como las **matemáticas** (creando una ecuación que modela el viaje del calor), si no que logró una forma de pensar que terminó siendo mucho más grande que este simple problema, pues es **transversal** a muchas otras cuestiones.

En este notebook realizaremos una **explicación profunda** y detallada de cada uno de los **elementos matemáticos** que conlleva la resolución de este problema, así como la consideración intuitiva del modelo que lo abstrae, logrando una solución que se ejecuta a partir de una herramienta de poder enorme: **la programación**, explicando cada algoritmo, método y herramienta utilizada, tanto para su **solución computacional** como para la optimización de esa respuesta en términos de *tiempo de ejecución*. Básicamente, entenderemos de forma **física** e **intuitiva** el significado de que el calor se *funda*, para traducirlo a un **modelo matemático** cuyas soluciones pueden hallarse haciendo uso de la **fuerza computacional**.

---

## **La Matemática que Necesitaremos.**

Antes de lanzarnos de lleno a **modelar cómo el calor se mueve**, tenemos que asegurarnos de tener las herramientas adecuadas para la **elaboración** y **comprensión del modelo**. Al igual que no se puede construir una casa sin saber utilizar un martillo, no se puede **modelar la realidad** sin tener claros ciertos conceptos clave.

Si queremos describir el viaje del **calor** sobre un objeto como una **barra metálica**, debemos tener en cuenta que el **fenómeno** de transmisión depende de dos cosas al mismo tiempo: la **posición** a lo largo de la barra y el **instante** en el que medimos la temperatura. Eso significa que nuestra **función** de temperatura **no puede** ser definida solo por una variable, en este caso es **indispensable** que se tengan en cuenta esos dos factores. Aquí necesitamos algo más elaborado: una función que recibe dos números **(dónde y cuándo)** y devuelve un número **(la temperatura)**. Existe una herramienta matemática que nos permite modelar este comportamiento, y son las **Funciones de Dos Variables**, una de las bases más fuertes del **Cálculo Multivariable**; a ese tipo de funciones, las podemos escribir de la siguiente manera:

$$f(x,t)$$

Donde la letra $f$ corresponde a la **función de temperatura**, la $x$ corresponde a la **posición de un punto de la barra**, y la $t$ se refiere a el **tiempo transcurrido**. En papel es simple, ¿no? Es simplemente agregarle una pequeña cosita más a las funciones simples que ya conocemos, sin embargo, detrás de esa simpleza, se esconde una complejidad **digna de admiración**: ahora tenemos herramientas para contestar cuestiones que involucren *el cambio* de esas dos variables, como por ejemplo, **¿cómo cambia la temperatura si dejo pasar un instante sin moverme de mi lugar?**, **¿Cómo cambia la temperatura si me muevo a cierta posición, sin haber pasado un instante?**, aquí es donde entra otra herramienta poderosísima: *la derivada parcial*.

Recordemos, brevemente, lo que significa el concepto de la **derivada**, evitando las formalidades, como el uso de límites y otras herramientas **formales** explicativas. Recordemos el concepto de la derivada de forma analítica. Si tenemos una función **univariable**, de toda la vida, definida como $f(x)$, si hallamos la **derivada** $f'(x)$ de esa función, vemos una información verdaderamente diciente sobre la función: **¿qué tan rápido está cambiando el valor de la función cuando movemos un poquito la $x$?** Si la derivada es grande y positiva, la función está subiendo bruscamente. Si es negativa, está bajando. Si es cero, está plana, al menos por un instante. Geométricamente, la derivada es la **pendiente** de la **recta tangente** a la curva en ese punto. Una **tangente** significa "que toca el objeto en un punto", como una regla que apoyamos suavemente sobre el **borde de una montaña**.

Eso es claro, **sencillo** y básico, pues son los fundamentos del **cálculo diferencial**, sin embargo, si tenemos en cuenta lo visto anteriormente, es completamente natural preguntarse: **ya, pero, ¿y si tenemos dos variables en una funcion?, ¿cómo se deriva en ese caso?** Ahí, la cosa se vuelve más interesante, pues no tenemos una única pendiente, tenemos dos:

1. Una que nos **indica** el cambio en la función si nos movemos en la **dirección de $x$** (dejando al tiempo $t$ fijo).
2. Otra que nos indica el cambio en la función si nos movemos en la **dirección de $t$** (dejando a la posición $x$ fija).

A eso, se le llaman **Derivadas Parciales**, una herramienta matemática que nos permite ver el cambio infinitamente instantáneo de una sola **variable** en una función multivariable, tratando como constantes a los otros elementos que definen a la función, si unimos la **derivada parcial** de cada una de las variables de una función en un **vector** $v$, lograremos un **vector** (flecha), que nos indica la dirección en la que la función más está creciendo; podemos definir a este tipo de derivadas de la siguiente manera **(formalmente)**:

$$\frac{\partial u}{\partial x}, \frac{\partial u}{\partial t}$$

El símbolo $\partial$ nos recuerda que **estamos derivando respecto a una variable** mientras tratamos a la otra como si fuera **constante**. Como si congeláramos el tiempo y solo miráramos cómo varía la **temperatura** a lo largo de la barra, o como si nos fijáramos en un punto fijo y observáramos cómo se calienta o se enfría con el paso de los segundos. Por ejemplo, definamos a la siguiente función bivariable, cuya gráfica debe representarse en un espacio de 3 dimensiones ($\mathbb{R}³$), tal que:

$$f(x, y) = 2x + y²$$

<center>
    <img src="plot1.png" width="650">
</center>

Como vemos, es una función que recibe **dos entradas** para dar una respuesta, pues debemos ingresar un valor para la $x$ y otro para la $y$. Entonces, en base a lo que hemos desarrollado anteriormente, **¿cómo derivamos entonces a esta función, si tiene dos variables?**, simple, utilizamos el conocimiento adquirido, y derivamos parcialmente. Primero, derivemos parcialmente la variable $x$, esto se hace, utilizando a cualquier variable $y$ como constante:

$$\frac{\partial}{\partial x} (2x + y²) = 2$$

Bien, tenemos una derivada parcial, ahora **apliquemos lo mismo**, pero para la variable $y$, derivando solo los términos que contengan a esa variable, y utilizando a los términos en $x$ como una constante:

$$\frac{\partial}{\partial y} (2x + y²) = 2y$$

Entonces, el **gradiente**, o el **vector** de dirección de la tasa de mayor **crecimiento** de la función en cualquier punto, denotado con el símbolo $\nabla$, corresponde a un vector en el que almacenamos las dos derivadas parciales calculadas previamente, tal que:

$$\nabla f(x, y) = (2, 2y)$$

Ahora, teniendo en cuenta estas ideas, podemos entrar a **pensar** en muchas más cosas, pues ya tenemos **ciertas herramientas** para poder responder preguntas importantes sobre el paso del calor. Como vimos, ya tenemos una función **bi-variable** que nos permite conocer la temperatura en base a la **posición** ($x$) en la barra y el **tiempo** ($t$), además, logramos visualizar una manera en la que podemos hallar el cambio que existe en la propagación de ese calor, utilizando el concepto de **derivadas parciales**. Sin embargo, todavía no estamos del todo listos para el trabajo, pues nos falta pensar en algo de vital importancia, que no estamos teniendo en cuenta; a ver, si la **temperatura** en un punto de la barra puede cambiar a medida que pasa el **tiempo**, o porque nos movemos hacia otro lugar (como definimos en nuestra **función de temperatura**), **¿cómo podemos relacionar esas dos formas de cambio?**, **¿Qué ecuación modela esta relación?**

La respuesta a esto es, **en esencia**, un concepto **matemático** de los más fascinantes e importantes en la historia, pues es uno de los motores por los cuales podemos **abstraer la realidad** y entenderla de forma matemática: **Las Ecuaciones Diferenciales Parciales**.

Sabemos que, de forma básica, una **ecuación normal**, de toda la vida, es una expresión en la que tenemos dos **miembros** (**expresiones**) separadas por un símbolo $=$, que contiene una o más incógnitas o (valores desconocidos); las **ecuaciones** son una forma básica de hallar respuestas a incógnitas o inquietudes con respecto a una expresión, y son herramientas que nos permiten hallar las **raíces** o **interceptos** de una parábola o cualquier polinomio, hallar múltiples incógnitas si hablamos de un **sistema de ecuaciones**, etc. Básicamente, estamos frente a una de las herramientas matemáticas más útiles en cualquier campo. Sin embargo, extraigamos la esencia de la ecuación y pensémosla en este plano: necesitamos hallar una relación que nos permita conocer cómo puede cambiar la **temperatura** con respecto al **tiempo** o la **posición**, es decir, aquí la incógnita no es una sola variable, es el comportamiento de una función entera.

Entonces pensemos, ¿qué tal si, en vez de que una **ecuación** resuelva el valor de una **incógnita única**, resuelve más bien la forma en la que una **función se comporta**?, **¿y si la incógnita no es solo una variable, si no que es una función entera?**. Esto, básicamente, es la intuición de las **ecuaciones diferenciales**, una ecuación en la que relacionamos una función desconocida, sus variables independientes y sus **derivadas**, no queremos hallar $x$, queremos hallar la forma de $f(x, t)$ tal que al derivarla de cierta manera, logremos una expresión en específico. Ahora, existen diferentes tipos de **ecuaciones diferenciales**, si la ecuación depende de **una sola variable** (como solo la posición o el tiempo), hablamos de **Ecuaciones Diferenciales Ordinarias (EDO)**, y la incógnita es $f(y)$ o $f(t)$, sus derivadas totales serían $f'(y)$ o $f''(t)$.

Pero cuando la función a encontrar se encuentra definida por **2 o más variables**, como es nuestro caso con la ecuación del calor, estamos hablando de **Ecuaciones Diferenciales Parciales (EDP)**; la ecuación aquí tratada es una de las más famosas y más **fáciles** de escribir, sin embargo, no siempre es una de las más faciles de resolver. Ahora bien, **¿por qué es importante hacer esta distinción?**, porque si una función tiene múltiples variables, puede cambiar en distintas direcciones de forma independiente para cada una. La **temperatura** puede variar con el **tiempo** sin variar con la **posición** (toda la barra se enfría **parejo**), o puede variar con la **posición** sin variar con el **tiempo** (un extremo **caliente** y otro **frío**, pero estable), o ambas cosas a la vez. Las derivadas parciales nos permiten hablar de cada tipo de cambio por separado.

**¡Perfecto!**, ya sabemos lo que es una **función multivariable**, una **derivada parcial** y  una **ecuación diferencial**, ya contamos con todas las herramientas para abordar un problema de este calibre, ahora podemos enfrentarnos a este dilema y utilizar nuestras bases para salir victoriosos, resolvamos la ecuación diferencial y salgamos de esto... muy bien, ¿cómo?:

- **¿De qué manera podemos utilizar este conocimiento para hallar solución a esto?**, y más importante...
- **¡¿Cómo modelamos la ecuación del calor?!**

---

## **Planteamiento del Modelo: Intuición Física para la Resolución.**

Muy bien, muy bien, las bases las **tenemos**, ahora pensemos en la **física** que debemos tener en cuenta para modelar en la matemática. Volvamos a la situación inicial, en la que encendemos la barra, como dijimos, el calor se *funde*, se *propaga*, para nada se queda quieto, lo que buscamos es modelar ese **movimiento**, pues en base a la empiria, y a lo que ocurre en esta situación, podemos afirmar que el calor tiende a **igualarse** naturalmente, como si odiara las diferencias, busca propagarse, lentamente, pero propagarse al fin y al cabo. Bien, **¿de qué depende la rapidéz de esto?** 

Inicialmente, de qué tan abrupta es la diferencia de temperatura; partimos de la premisa que nos dice que si situamos **dos puntos en la barra**, uno muy **caliente** y el otro **contrariamente frío**, el calor se propagará muy rápido por la necesidad del mismo de expandirse, en cambio, si situamos **dos puntos** nuevamente, uno muy **caliente**, y el otro un poco **menos caliente** que el anterior, la propagación será más lenta.

Segundo punto, el material, pues hay materiales, como el **cobre (Cu)**, que se calientan y enfrían rápidamente, o hay otros muchos como la **madera** que cambian su temperatura de forma muy lenta. Dependiendo del material, podemos definir una *"personalidad térmica"*, o más formalmente conocida como una *difusividad térmica* (cuyo nombre será $\alpha$ en este caso).

En **tercera instancia**, otro factor que influye mucho en la **termodinámica** es la forma en la que el calor se distribuye alrededor de cada punto, pues no es lo mismo que un **punto medio** esté muy caliente y sus **extremos fríos**, a que esté muy frío y sus **extremos calientes**, y esto es un punto crucial que nos dará la respuesta del modelado. Dejemos todo lo que hemos desarrollado anteriormente y vamonos a algo mucho más simple, retrocedamos al **cálculo diferencial** de *"coquito"*.

El **cálculo diferencial** corresponde a la base de cada una de las expresiones que hemos usado aquí, sin embargo, es necesario retroceder para precisar una idea que nos permitirá tener una **mayor claridad**. El punto central de esta vertiente del cálculo, es la capacidad de ver el cambio instantáneo de cualquier función, ¿no?, Claro, eso es precisamente lo que hacemos al **derivar**, pero, ¿para qué nos sirve eso?

Fácil, la **derivada**, junto a otra herramienta increíble como lo son los límites, nos permiten, básicamente, analizar el **comportamiento** de cualquier función, y a partir de eso graficarla; por su parte, los límites nos permiten ver discontinuidades, tendencias al **infinito** de una función, encontrar particularidades en tramos específicos. Por otra parte, la *derivada* nos hace poder analizar **el cambio** de una función, ¿cómo? A partir de dos elementos básicos que estabamos dejando de lado: los **criterios de derivación**.

Cuando decimos que la **derivada** nos permite conocer el cambio inmediato de una función, estamos diciendo ímplicitamente, que a partir de ella, podemos encontrar los puntos específicos en los que **crece** o **decrece**, eso significa, precisamente, el **cambio**. Ahora bien, **¿cómo se hallan esos puntos?** Simple, a partir del ***Criterio de la Primera Derivada***. Fijemos una función $f(x)$ tal que:

$$f(x) = 7x³ + 9x² - 2x - 25$$

Sabemos (si hemos pasado por un curso base de cálculo diferencial) que el ***criterio de la primera derivada*** nos dice que podemos hallar los puntos críticos de **crecimiento** o **decrecimiento** de una función, si la derivamos e igualamos su derivada a $0$, ¿por qué es así? Intuición matemática, si la **recta tangente** a un punto de la función tiene una pendiente de $0$,  es porque hemos encontrado o un **máximo local** o un **mínimo local**. A esto se le llaman puntos críticos y son los puntos de crecimiento o decrecimiento. Hagámoslo con la función planteada arriba, y grafiquemoslo:

<center>
    <img src="derivative_criteria_of_a_function.png" width="650">
</center>

$$f'(x) = 21x² + 18x -2  \ ; \ 21x² + 18x - 2 = 0$$
$$x_1 = \frac{-9 - \sqrt{123}}{21}, x_2 = \frac{-9 + \sqrt{123}}{21}$$

Esos son los puntos críticos de la función, con $x_1$ siendo la **coordenada** en $x$ del punto máximo, y $x_2$ la coordenada en $x$ del **mínimo local**. Con esto, hallamos los puntos de crecimiento y decrecimiento de la función, sin embargo, hay algo más allá: un ***Criterio de la Segunda Derivada***, que nos dice que si **derivamos** por segunda vez la función, podemos hallar los puntos de **inflexión** de una función, es decir, los puntos en los que la función **cambia de concavidad**. Para hallarlos, derivamos nuevamente la función, e igualamos a $0$:

$$f''(x) = 42x + 18 \ ; \ 42x + 18 = 0$$
$$x = -\frac{18}{42} = -\frac{3}{7}$$

La función **cambia de concavidad** en la coordenada $-\frac{3}{7}$ del eje $x$, perfecto, analizamos la **concavidad** de una función. Ahora bien, **¿qué tiene que ver eso con la ecuación del calor?**, absolutamente todo, porque al analizar la **concavidad** o **encorvamiento** de la función de perfil térmico en base a la posición, logramos definir el cambio de la temperatura con respecto al tiempo y a la posición. Esto que planteamos responde a lo que establecimos anteriormente, si hay un cambio muy abrupto de temperatura de un punto a otro, se va a transmitir **más rápido el calor**, si hay un **cambio** poco notorio, el calor se va a transmitir **más lentamente**, y definimos el cambio de un punto a otro, con la concavidad (o segunda derivada) de la función. Así pues:

> *"La rapidez con que cambia la temperatura en un punto es proporcional a qué tan "encorvado" está el perfil de temperatura alrededor de ese punto, y esa proporcionalidad tiene en cuenta la difusividad térmica del material."*

Teniendo clara esta idea, tenemos la solución, podemos plantear la **ecuación diferencial parcial** que nos puede dar respuesta a este problema; pensémosla así:

$$\frac{\partial u}{\partial t} = \alpha \frac{\partial ² u}{\partial x²}$$

Donde $u$ es una función $u(x, t)$ que modela la diferencia de **temperatura** con respecto a el tiempo y la **posición en la barra**, esa ecuación diferencial parcial modela lo que hemos descrito anteriormente, porque al escribirla, estamos diciendo que la velocidad con que **cambia la temperatura** en un punto fijo **(derivada temporal: $\partial t$)** es proporcional a la curvatura del perfil de temperatura alrededor de ese punto **(segunda derivada espacial: $\partial x²$)**. ¡Perfecto!, podemos avanzar, tenemos el problema planteado, ¡solo es solucionar esa ecuación y estamos listos!, ahora... ¿cómo?

Si **damos un vistazo a la literatura**, existen múltiples métodos para encontrarle solución a este tipo de ecuaciones, podemos utilizar uno simple, **manteniendo consideraciones clave**, por ejemplo, en este caso, podemos asumir que existe una solución separable a la ecuación, tal que $u(x, t) = X(x) * T(t)$, en otras palabras, reescribimos la función como el producto entre una función que solo tiene en cuenta la **posición** ($x$) y otra que solo tiene en cuenta el **tiempo** ($t$), ahora llevamos esto a la ecuación diferencial original, separamos variables y logramos dos **Ecuaciones Diferenciales Ordinarias**, una para $X(x)$ y otra para $T(t)$, ecuaciones las cuales se solucionan de forma separada, utilizando sustituciones trigonométricas, exponenciales, u otras, dependiendo del entorno de la ecuación.

Luego, es importante aplicar las **condiciones de contorno** para determinar soluciones válidas, es decir, la solución de la ecuación puede variar dependiendo de condiciones como la temperatura en los extremos y en el centro; aplicamos condiciones iniciales y obtenemos la respuesta. 

Este es el algoritmo a seguir que enseñan en cualquier curso serio de **Ecuaciones Diferenciales**, un método contínuo, en el que se utiliza álgebra simbólica, sustituciones, entre otras muchas técnicas numéricas para el cálculo y resolución de estas ecuaciones. Esta es una de las formas más reconocidad de resolución, pues es una forma de resolver el problema *"en papel"*, sin embargo, si queremos optimizar la resolución de un problema, queremos computar ese problema, automatizarlo utilizando **herramientas computacionales**; ante esto, pueden surgir dudas válidas como:

- **¿Se puede computar una resolución de este tipo para estas ecuaciones?**
- **¿Es esta una forma eficiente de resolverlo?**
- **¿Se puede mejorar esta resolución?**

Las respuestas a estas preguntas son, cuanto menos, **desalentadoras**, pues no brindan las respuestas que quisieramos, pues existe un problema muy grande con esta forma inicial de resolución que propusimos. Primero, **cuestionémonos** algo, pues podemos **solucionarlo** de una forma relativamente fácil si lo hacemos en *"papel"*, sin embargo, **¿existe alguna forma de computar un algoritmo con tantas consideraciones?**; sí existe, se llama **álgebra computacional**, sin embargo, **no es una opción eficiente**, pues este tipo de ecuaciones requieren de técnicas algbráicas muy complejas, pensemos en que el manejo de matemática simbólica requiere de identidades, sustituciones que son muy pesadas en términos de memoria; para hacer transformaciones simples, existe documentación enorme, y una cantidad de reglas, que hacen que no sea una forma óptima de resolver el problema.

Además, este método funciona muy bonito para casos muy específicos: si la **barra es uniforme**, si los **extremos están a temperatura fija**, si la condición inicial tiene forma de **seno**. Pero si cambiamos un poquito alguna de esas condiciones, la solución analítica se vuelve **extremadamente complicada**. Y si las cambiamos mucho, directamente no existe una expresión cerrada que podamos escribir con funciones elementales.

Entonces, **¿qué hacemos?**, **¿Cómo podemos aprovechar las capacidades computacionales?**, **¿Hay alguna alternativa ante esto?**

---

## **Cuando las Artesanías no Alcanzan: Introducción al Método de Diferencias Finitas**

Pensemos en esto: la barra es un objeto **contínuo**, tiene infinitos puntos. Sin embargo, nosotros no necesitamos saber la **temperatura** en todos los puntos, necesitamos conocerla en **suficientes puntos** como para entender qué está pasando. Es, **análogamente**, como cuando dibujamos una curva en un **papel**. No trazamos cada punto infinito que la compone. Tomamos algunos puntos clave y los unimos con trazos suaves. Si los puntos están bien elegidos, la curva se ve igual.

Hagamos lo mismo con nuestra **barra**, en lugar de tratarla como un contínuo, partámosla en pedazos más pequeños. Dividámosla en un número definido de puntos **equiespaciados** $\Delta x$. Al primero, lo llamamos $x_0$, al segundo $x_1$ y así sucesivamente hasta llegar hasta $x_{n-1}$. Hagamos lo mismo con el tiempo, nuevamente, no necesitamos saber qué pasa en cada instante. Elegimos una cantidad **finita** de pasos de tiempo, separados por un intervalo $\Delta t$ equidistante para cada instante.

Así pues, en lugar de tener una función contínua $u(x, t)$, tenemos un **conjunto de valores** $u_i^n$, donde $i$ es el **punto de la barra que observamos** y $n$ es el punto de tiempo en el que nos encontramos. La disyuntiva que nos compete es **¿cómo actualizamos estos valores?** Es decir, si conocemos la **temperatura** en todos los puntos en el tiempo $n$, **¿cómo calculamos la temperatura en el tiempo $n+1$?**

Recordemos que la **ecuación de calor planteada** nos habla de dos cosas: una derivada temporal ($\frac{\partial u}{\partial t}$) que nos habla del cambio de la **temperatura** con respecto al tiempo, y una derivada **espacial** ($\frac{\partial ² u}{\partial x²}$) que es la curvatura del perfil de temperatura. Pero hay un problema, dividimos la barra en puntos, ya no tenemos funciones continuas, tenemos puntos discretos. Así que **tenemos** que aproximar ambas **derivadas** usando solo los valores en esos puntos, **¿cómo lo hacemos?**

Pensemos en un **punto fijo** de la barra, generalicemos el punto $x_i$​. En un tiempo cualquiera $t_n$, su temperatura es $u_i^n$​; entonces, en el siguiente paso de tiempo $t_{n + 1}$​, su temperatura será $u_i^{n+1}$​. La derivada temporal nos dice la velocidad con la que cambió la temperatura; en un mundo contínuo, sería el límite cuando el intervalo de tiempo tiende a cero (por la definición de **derivada**). Pero nosotros tenemos un intervalo de tiempo pequeño pero finito $\Delta t$.

Lo más razonable y sencillo sería restar la **temperatura inicial**,​ a la **temperatura final**, y dividir entre el intervalo de tiempo; al hacer eso, estamos hallando el cambio de la **temperatura** con respecto al **tiempo**. Planteamos, entonces:

$$\frac{\partial u}{\partial t} \approx \frac{u_i^{n+1} - u_i^n}{\Delta t}$$

Esto se llama **diferencia hacia adelante en el tiempo**, miramos hacia el paso siguiente y hacia atrás (el paso actual), y vemos la diferencia, sencillo. Ahora, para **aproximar** la curvatura, la **segunda derivada espacial** nos dice cómo cambia la **pendiente** a lo largo de la barra. Esto podemos aproximarlo mirando al punto actual y sus dos vecinos: el de la izquierda y el de la derecha. La fórmula para hallar esto, que sale de la misma lógica anterior, es:

$$\frac{\partial ² u}{\partial x²} \approx \frac{u_{i + 1}^n - 2u_i^n + u_{i-1}^n}{\Delta x²}$$

Notar dos cosas: esta expresión no usa **ningún punto** que no sea el **actual** y sus **vecinos inmediatos**, es local. Eso será clave más adelante cuando queramos paralelizar. Además, notar que es un resultado aproximado, es decir, al discretizar el problema hacemos *aproximaciones*, no hallamos respuestas exactas. Ahora, si juntamos piezas y reemplazamos lo encontrado en la **función de calor** original, obtenemos:

$$\frac{u_i^{n+1} - u_i^n}{\Delta t} = \alpha \frac{u_{i + 1}^n - 2u_i^n + u_{i-1}^n}{\Delta x²}$$

Y si despejamos $u_i^{n+1}$, que es lo que queremos calcular (la temperatura en el **futuro**), nos queda:

$$u_i^{n+1} = \alpha \frac{\Delta t}{\Delta x²} (u_{i + 1}^n - 2u_i^n + u_{i-1}^n)  + u_i^n$$

Y esto es, en definitiva, el **método explicito** de ***Diferencias Finitas*** para la ecuación de calor. Con esto  logramos algo gigante, y es que pasamos de tener que resolver una **ecuación diferencial parcial**, una operación complicada, que requiere álgebra simbólica, sustituciones, integrales, etc., a tener una fórmula de **actualización** que solo usa las operaciones matemáticas básicas: **sumas**, **restas**, **multiplicaciones** y **divisiones**. Ahora, es un problema que **se puede resolver con un cálculo simple**, cualquiera puede escribirlo en una calculadora y recibir un resultado en cierto tiempo y una calculadora lo resuelve en **fracciones de segundo**; este es, realmente, una optimización fascinante a el problema.

Ahora, notemos algo, y es que cada punto $u_i^{n + 1}$ se calcula usando solo valores del paso de tiempo anterior $n$, no necesitamos saber nada de otros **puntos** en el futuro. Eso significa que podemos calcular todos los puntos del siguiente paso de tiempo de forma **independiente**, esto nos señala una cosa, y es que al ser una **resolución** de esta índole, podemos **paralelizar** estas operaciones en diferentes **hilos de ejecución** al modelar esta ecuación **computacionalmente**. Pero no nos adelantemos, eso lo veremos más adelante. Primero, necesitamos probar que nuestra aproximación sea **correcta**, ahora, **¿cómo lo confirmamos?**

Por suerte, para el caso **particular** que vamos a resolver (una barra de longitud $L$, con extremos a temperatura cero y una **condición inicial sinusoidal**), existe una solución analítica. Fourier mismo la encontró, y es la siguiente:

$$u(x, t) = \sin\left(\frac{\pi x}{L}\right) e^{- \frac{\pi ² \alpha}{L²}t}$$

Si la interpretamos, vemos que con la parte de $\sin(\frac{\pi x}{L})$, definimos la forma inicial de la **temperatura** a lo largo de la barra, es una **campana suave**, con 0 en los extremos y una máxima en el centro. La parte de $e^{- \frac{\pi ² \alpha}{L²}t}$ es una función exponencial decreciente, con ella modelamos que, a medida de que el tiempo pasa, este factor se hace **cada vez más pequeño**, el calor se **disipa** y la barra se enfría. Esta solución es exacta, no una aproximación, y por eso es justo lo que necesitamos, porque podemos ejecutar nuestro **método numérico**, obtener un resultado estimado, y comparar con el resultado real; si el error es pequeño, confiamos en este método, si es grande, algo hay que corregir.

Ahora, ya tenemos la **fórmula**, sabemos que la podemos **computar** de forma simple, pues solo usamos operaciones **aritméticas sencillas**, además, sabemos que existe una solución exacta y contrastada con la que podemos comparar, entonces, **¿como extrapolamos esto a un programa en C?**

Necesitamos varias cosas, entre ellas, definir a la barra como un **conjunto de puntos**, escribir un bucle que actualice cada punto utilizando **nuestra fórmula**, aplicar las **condiciones de contorno** y finalmente, iterar para cada **paso de tiempo** comparando con la solución exacta. Parece mucho, pero veremos que el código es sorprendentemente corto. La magia no está en la cantidad de líneas, sino en la claridad de las ideas, observemos la implementación:

In [5]:
//%ldflags: -lm
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include <omp.h>

#define PI 3.14159265358979323846

#define L 1.0
#define alpha 0.01
#define T_max 0.5
#define N_x 100
#define N_t 1000

void inicializar_condicion_inicial(double* u, int n_x, double dx) {
    for (int i = 0; i < n_x; i++) {
        double x = i * dx;
        u[i] = sin(PI * x / L);
    }
}

void aplicar_condiciones_frontera(double* u, int n_x) {
    u[0] = 0.0;
    u[n_x - 1] = 0.0;
}

void resolver_ecuacion_calor(double* u, double* u_nueva, int n_x, int n_t, double dx, double dt, double alpha_value) {
    double coef = alpha_value * dt / (dx * dx);
    
    for (int t = 0; t < n_t; t++) {
        for (int i = 1; i < n_x - 1; i++) {
            u_nueva[i] = u[i] + coef * (u[i+1] - 2 * u[i] + u[i-1]);
        }
        
        aplicar_condiciones_frontera(u_nueva, n_x);
        
        memcpy(u, u_nueva, n_x * sizeof(double));
    }
}

double solucion_exacta(double x, double t) {
    return sin(PI * x / L) * exp(-alpha * (PI * PI) / (L * L) * t);
}

int main() {
    double dx = L / (N_x - 1);
    double dt = T_max / N_t;
    double coef = alpha * dt / (dx * dx);
    
    printf("Coeficiente de estabilidad: %f\n", coef);

    if (coef > 0.5) {
        printf("Advertencia: el metodo puede ser inestable.\n");
    }
    
    double* u = (double*) malloc(N_x * sizeof(double));
    double* u_nueva = (double*) malloc(N_x * sizeof(double));
    
    inicializar_condicion_inicial(u, N_x, dx);
    aplicar_condiciones_frontera(u, N_x);
    
    double start_time = omp_get_wtime();

    resolver_ecuacion_calor(u, u_nueva, N_x, N_t, dx, dt, alpha);

    double end_time = omp_get_wtime();

    double elapsed_ms = (end_time - start_time) * 1000.0;
    
    printf("Tiempo de ejecucion: %f ms\n", elapsed_ms);
    
    double error_max = 0.0;

    for (int i = 0; i < N_x; i++) {
        double x = i * dx;
        double exacta = solucion_exacta(x, T_max);
        double error = fabs(u[i] - exacta);

        if (error > error_max) {
            error_max = error;
        }
    }

    printf("Error maximo al tiempo T = %f: %e\n", T_max, error_max);
    
    FILE *file = fopen("resultados.dat", "w");

    fprintf(file, "x u_numerica u_exacta error\n");

    for (int i = 0; i < N_x; i++) {
        double x = i * dx;
        double exacta = solucion_exacta(x, T_max);
        double error = fabs(u[i] - exacta);

        fprintf(file, "%f %f %f %f\n", x, u[i], exacta, error);
    }

    fclose(file);

    free(u);
    free(u_nueva);
    
    return 0;
}

Coeficiente de estabilidad: 0.049005
Tiempo de ejecucion: 1.812834 ms
Error maximo al tiempo T = 0.500000: 2.782417e-06


Primero, definimos las cabeceras de compilación `<stdio.h>`, `<stdlib.h>` (para el manejo de arreglos **dinámicos** en memoria), `<math.h>` (para el uso de funciones matemáticas específicas), y `<string.h>` (para utilizar `memcpy()`, una función que permite la copia fácil de arreglos), `omp.h` (para el uso de `omp_get_wtime()`, que nos permite medir tiempos de ejecución). 

Luego definimos $\pi$ con una precisión bastante alta utilizando `#define PI 3.14159265358979323846`, y las constantes del problema:

- `#define L 1.0` es la **longitud de la barra**, un metro. 
- `#define alpha 0.01` es la **difusividad térmica** del material.
- `#define T_max 0.5` es el **tiempo total** que vamos a simular, medio segundo. 
- `#define N_x 100` es la **cantidad de puntos en los que dividimos la barra**. 
- `#define N_t 1000` es la **cantidad de pasos de tiempo** que vamos a dar.

Todas estas constantes están **al principio** porque si queremos cambiar algún **parámetro** (una barra más larga, un material diferente, más precisión), solo **modificamos** estos números en un solo lugar y todo el programa se ajusta automáticamente.

Después de las constantes, definimos la función `inicializar_condicion_inicial()`. Recibe el arreglo `u`, la cantidad de puntos `n_x` y el paso espacial `dx`. Dentro de ella, hay un bucle `for` que recorre todos los puntos de la barra. Para cada punto, calcula su posición `x = i * dx` y le asigna la temperatura inicial `sin(PI * x / L)`. **¿Por qué un seno?** Porque, como lo dijimos antes, es una condición inicial que tiene una solución analítica exacta, y podemos comparar nuestro resultado con esa solución.

Luego definimos `aplicar_condiciones_frontera()`. Esta función recibe el arreglo `u` y la cantidad de puntos `n_x`, y simplemente asigna cero al primer punto `(u[0])` y al último punto `(u[n_x - 1])`. Esto representa físicamente que los extremos de la barra se mantienen a temperatura constante de cero grados durante toda la simulación. Sin estas condiciones, la ecuación diferencial no tendría una única solución.

Ahora viene la función más importante: `resolver_ecuacion_calor()`. Recibe los dos arreglos (`u` y `u_nueva`), la cantidad de puntos `n_x`, la cantidad de pasos de tiempo `n_t`, el paso espacial `dx`, el paso temporal `dt`, y la difusividad térmica `alpha_value`. Lo primero que hace la función es calcular el coeficiente de estabilidad `coef = alpha_value * dt / (dx * dx)`. Este número es clave: si es mayor a $0.5$, el método explota numéricamente y los resultados no tienen sentido físico. Por eso en main lo imprimimos y mostramos una advertencia si supera el límite.

Dentro de la función, hay un bucle externo que se repite `n_t` veces, una por cada paso de tiempo. En cada paso de tiempo, hay un bucle interno que recorre todos los puntos interiores de la barra, desde `i = 1` hasta `i = n_x - 2`. Los extremos `i = 0` e `i = n_x - 1` no se actualizan con la fórmula, porque ya los fijamos con las condiciones de contorno. Para cada punto interior, aplicamos la fórmula de diferencias finitas: 

`u_nueva[i] = u[i] + coef * (u[i+1] - 2*u[i] + u[i-1])`. 

Esto nos da la temperatura en ese punto en el siguiente paso de tiempo. Después de calcular todos los puntos nuevos, aplicamos las condiciones de contorno a u_nueva para asegurarnos de que los extremos sigan siendo cero. Finalmente, usamos `memcpy()` para copiar los valores de `u_nueva` a `u` de una sola vez, de forma eficiente. Así, en la próxima iteración del bucle temporal, u contiene la temperatura actualizada.

Después definimos `solucion_exacta()`, una función que recibe una posición $x$ y un tiempo $t$, y devuelve el valor de la solución analítica `sin(PI * x / L) * exp(-alpha * (PI * PI) / (L * L) * t)`. Esta función la usaremos al final para comparar nuestro resultado **numérico** con el valor real y calcular el error.

En el `main()`, lo primero que hacemos es calcular `dx`, `dt` y `coef` a partir de las constantes; 

- `dx = L / (N_x - 1)`. ¿Por qué N_x - 1? Porque si tenemos `N_x` puntos, hay `N_x - 1` intervalos entre ellos. Por ejemplo, si la barra mide 1 metro y tenemos 100 puntos, la distancia entre punto y punto es 1 metro dividido por 99 intervalos. 
- `dt = T_max / N_t`, que es simplemente el tiempo total dividido por la cantidad de pasos. Mostramos el coeficiente en pantalla y verificamos si es mayor a 0.5. Si lo es, mostramos una advertencia. Finalmente, liberamos la memoria que reservamos con `free(u)` y `free(u_nueva)`.  Esto es importante porque si no liberamos la memoria, el programa la seguiría ocupando hasta que termine, y en programas grandes o que se ejecutan muchas veces, eso puede causar problemas de memoria.

Ahora, **¿nuestro programa es matemáticamente válido?**, Probémoslo, si lo ejecutamos, obtenemos tres líneas de **información directa**. La primera fue el **coeficiente de estabilidad**, que fue de `0.049005`. Recordemos que la condición que nos asegura que el método no explote es que este número **sea menor o igual a `0.5`**. Nuestro valor es casi **diez veces** más pequeño. **¿Qué significa esto en términos prácticos?** Significa que estamos operando en una región extremadamente segura. No hay riesgo de oscilaciones numéricas, no van a **aparecer temperaturas disparatadas que suban y bajen sin control**. Podríamos incluso aumentar el paso de tiempo o reducir el paso espacial sin que el método se vuelva inestable. Pero por ahora, estamos sobradamente protegidos.

El tiempo de ejecución fue de `1.598626` milisegundos. Apenas una milésima y media de segundo. Hagamos cuentas rápidas: tenemos `N_x = 100` puntos espaciales y `N_t = 1000` pasos de tiempo. Eso significa que nuestro programa realizó `100 * 1000 = 100000` actualizaciones de temperatura. Cada actualización implica unas pocas sumas, restas y multiplicaciones. Cien mil operaciones en menos de dos milisegundos. Eso es ridículamente rápido. Un procesador moderno barre estas operaciones sin esfuerzo. Pero atención: esto es solo con `100` puntos. Si quisiéramos simular una barra con mucha mayor resolución, digamos `10000` puntos, el número de operaciones saltaría a `10000 * 1000 = 10^7`, y el tiempo crecería proporcionalmente. Ahí es donde la optimización empieza a importar.

El error máximo al tiempo `T = 0.5` segundos fue de `2.782417e-06`. Eso es aproximadamente `0.00000278` grados. Un error del orden de la millonésima parte de un grado. Para ponerlo en perspectiva: un termómetro común tiene una precisión de décimas de grado. Un termómetro de laboratorio de alta calidad llega a milésimas. Nuestro error está tres órdenes de magnitud por debajo de eso. En la práctica, para cualquier aplicación de ingeniería real, esta precisión es más que suficiente. Nuestro método numérico está haciendo su trabajo con una fidelidad considerable.

Ahora, en el lugar en donde se concentra la mayor cantidad del **error**, es en los puntos situados cercanos a el centro de la barra, pues es ahí donde la función tiene una mayor **curvatura**, la concavidad es mucho más **pronunciada** y la **aproximación numérica** se nota un poco más, sin embargo, sigue siendo un error casi **despreciable**. Ahora, observemos la siguiente gráfica, en la que comparamos el perfil de temperatura al tiempo de `T = 0.5` segundos para ambas soluciones, la **solución numérica** y la **solución** exacta:

<center>
    <img src="temperatura_final.png" width="650">
</center>

De esta forma, comprobamos visualmente el **funcionamiento** de nuestro método, a simple vista, las dos curvas son **indistinguibles**. Se montan una encima de la otra como si fueran la misma línea. Esto es una señal clarísima de que nuestro método funciona, y que lo hace muy bien. La curva tiene forma de **campana**. Comienza en cero en el **extremo izquierdo**, **sube suavemente**, alcanza un pico en el **centro de la barra**, y luego baja simétricamente hasta cero en el extremo derecho. Esta forma no es casual. La condición inicial era un **seno**, que también tiene forma de campana. Pero con el paso del tiempo, el calor se ha difundido. El pico se ha achatado un poco. La curva se ha vuelto más ancha y más baja. Eso es exactamente lo que esperaríamos físicamente: el calor que estaba concentrado en el centro se ha ido esparciendo hacia los bordes, y como los bordes se mantienen a temperatura cero, el calor termina escapando de la barra. En conjunto, tanto los números como la gráfica nos confirman una sola cosa: **nuestra implementación del método de diferencias finitas para la ecuación del calor es correcta**, es **precisa** y es **eficiente**. El error es del orden de la **millonésima**. El tiempo de ejecución es de apenas milisegundos.

Ahora, una pregunta importante: **¿este código es óptimo?** En términos de complejidad algorítmica, tenemos dos bucles anidados: uno sobre los pasos de tiempo (que se repite `n_t` veces) y otro sobre los **puntos espaciales** (que se repite `n_x` veces). La complejidad temporal es $O(N_t \cdot N_x)$, es decir, crece linealmente con la cantidad de pasos de tiempo y con la cantidad de puntos espaciales. Si duplicamos la resolución espacial o temporal, el tiempo de ejecución se duplica. Eso es razonable y no se puede mejorar mucho porque, en el fondo, tenemos que actualizar cada punto en cada paso de tiempo.

En términos de complejidad espacial, usamos dos arreglos de tamaño `n_x`, más unas pocas variables locales. La memoria que ocupa el programa es $O(N_x)$, es decir, crece linealmente con la cantidad de puntos espaciales. Si queremos más precisión espacial (más puntos), necesitamos más memoria, no hay forma de evitar eso.

Ahora, hay algo que podemos mejorar: el bucle espacial, el que recorre los puntos interiores de la barra, es completamente paralelizable. **¿Por qué?** Porque cada punto `i` solo necesita los valores de `u[i-1]`, `u[i]` y `u[i+1]`, que ya están disponibles y no cambian mientras se calcula el mismo paso de tiempo. No hay dependencias entre los cálculos de diferentes puntos. Eso significa que podemos dividir el trabajo entre varios hilos y que cada hilo calcule un subconjunto de puntos. La única forma de lograr lo anterior, es paralelizando el código para lograr mejoras en el rendimiento... ¿o no?.

## **La Paralelización Como... ¿Alternativa de Mejora?**

Tenemos un código limpio, funcional y **óptimo**, pero a pesar de esta gran implementación, podemos mejorarlo y hacerlo incluso más útil; si en este momento implementamos una **gran cantidad** de divisiones, es decir de un orden mayor al de $10⁷$, por ejemplo, va a ser una demanda computacional mucho mayor, pues el orden de operaciones a realizar creceria **considerablemente**, sin embargo, podemos mitigar eso utilizando una herramienta mencionada anteriormente: **La Paralelización**. 

Podemos hacer que este código sea capáz de aguantar volúmenes de iteraciones enormes, **simplemente** realizando una **división en hilos** para ciertas operaciones. Observa la **implementación** de nuesto método de solución añadiendo **directivas de paralelización**:

In [6]:
//%ldflags: -lm -fopenmp
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include <omp.h>

#define PI 3.14159265358979323846

#define L 1.0
#define alpha 0.01
#define T_max 0.5
#define N_x 100
#define N_t 1000

void inicializar_condicion_inicial(double* u, int n_x, double dx) {
    for (int i = 0; i < n_x; i++) {
        double x = i * dx;
        u[i] = sin(PI * x / L);
    }
}

void aplicar_condiciones_frontera(double* u, int n_x) {
    u[0] = 0.0;
    u[n_x - 1] = 0.0;
}

void resolver_ecuacion_calor_paralelo(double* u, double* u_nueva, int n_x, int n_t, double dx, double dt, double alpha_value) {
    double coef = alpha_value * dt / (dx * dx);
    
    for (int t = 0; t < n_t; t++) {
        #pragma omp parallel for
        for (int i = 1; i < n_x - 1; i++) {
            u_nueva[i] = u[i] + coef * (u[i+1] - 2 * u[i] + u[i-1]);
        }
        
        aplicar_condiciones_frontera(u_nueva, n_x);
        
        memcpy(u, u_nueva, n_x * sizeof(double));
    }
}

double solucion_exacta(double x, double t) {
    return sin(PI * x / L) * exp(-alpha * (PI * PI) / (L * L) * t);
}

int main() {
    double dx = L / (N_x - 1);
    double dt = T_max / N_t;
    double coef = alpha * dt / (dx * dx);
    
    printf("Coeficiente de estabilidad: %f\n", coef);

    if (coef > 0.5) {
        printf("Advertencia: el metodo puede ser inestable.\n");
    }
    
    double* u = (double*) malloc(N_x * sizeof(double));
    double* u_nueva = (double*) malloc(N_x * sizeof(double));
    
    inicializar_condicion_inicial(u, N_x, dx);
    aplicar_condiciones_frontera(u, N_x);
    
    double start_time = omp_get_wtime();
    resolver_ecuacion_calor_paralelo(u, u_nueva, N_x, N_t, dx, dt, alpha);
    double end_time = omp_get_wtime();

    double elapsed_ms = (end_time - start_time) * 1000.0;
    
    printf("Tiempo de ejecucion (código en paralelo): %f ms\n", elapsed_ms);
    
    double error_max = 0.0;

    for (int i = 0; i < N_x; i++) {
        double x = i * dx;
        double exacta = solucion_exacta(x, T_max);
        double error = fabs(u[i] - exacta);

        if (error > error_max) {
            error_max = error;
        }
    }

    printf("Error maximo al tiempo T = %f: %e\n", T_max, error_max);
    
    FILE *file = fopen("resultados_paralelo.dat", "w");

    fprintf(file, "x u_numerica u_exacta error\n");

    for (int i = 0; i < N_x; i++) {
        double x = i * dx;
        double exacta = solucion_exacta(x, T_max);
        double error = fabs(u[i] - exacta);

        fprintf(file, "%f %f %f %f\n", x, u[i], exacta, error);
    }

    fclose(file);

    free(u);
    free(u_nueva);
    
    return 0;
}

Coeficiente de estabilidad: 0.049005
Tiempo de ejecucion (código en paralelo): 81.478857 ms
Error maximo al tiempo T = 0.500000: 2.782417e-06


La única diferencia entre el código serial y el paralelo está dentro de la función `resolver_ecuacion_calor`. En la versión paralela, aparece esta línea justo antes del bucle `for` que recorre los puntos de la barra:

`#pragma omp parallel for`

Esta directiva le dice al compilador: **"toma el bucle que sigue y ejecuta sus iteraciones en paralelo usando múltiples hilos"**. OpenMP se encarga automáticamente de:

- Dividir el rango de iteraciones (desde `i = 1` hasta `i = n_x - 2`) entre los hilos disponibles.
- Asignar a cada hilo un bloque de puntos para que los calcule simultáneamente.
- Sincronizar los hilos al final del bucle (esperar a que todos terminen antes de continuar).

El resto del código es **idéntico**: las condiciones de frontera, la copia con `memcpy`, la inicialización, la solución exacta. Todo sigue igual.

Ejecutamos ambos códigos con `N_x = 100` y `N_t = 1000`. Los resultados fueron:

<div align = "center">

| Versión | Tiempo de ejecución | Error máximo |
|---------|---------------------|--------------|
| Serial  | 1.81 ms             | 2.782417e-06     |
| Paralelo| 81.478857 ms        | 2.782417e-06    |

</div>


**El error es idéntico**, porque el cálculo numérico no cambia. Pero el tiempo de ejecución es **abismalmente diferente**: el paralelo tardó más de 80 milisegundos, mientras que el serial tardó solo 1.81 milisegundos. La versión paralela fue **más de 80 veces más lenta**.

Parece contradictorio: si dividimos el trabajo entre varios hilos, debería ejecutarse más rápido. Pero hay un costo oculto que muchos olvidan: el **overhead**.

OpenMP no es mágico. Cuando el programa llega a `#pragma omp parallel for`, deben ocurrir varias cosas antes de que los hilos realmente comiencen a trabajar:

1. **Creación o activación de hilos:** OpenMP debe crear (o despertar) múltiples hilos. Esto toma tiempo.
2. **Distribución de iteraciones:** El compilador genera código para decidir qué hilo calcula qué puntos. Esto también toma tiempo.
3. **Sincronización:** Al final del bucle, todos los hilos deben esperar al más lento. Hay una barrera implícita.
4. **Destrucción o suspensión de hilos:** Al salir del bucle, los hilos se duermen o destruyen.

Ahora, multiplica estos costos por la cantidad de pasos de tiempo. En nuestro caso, `N_t = 10000`. Eso significa que **el overhead de OpenMP ocurre 10000 veces**, una por cada paso de tiempo.

¿Cuántas operaciones útiles hay por paso de tiempo? Solo `N_x - 2 ≈ 500` operaciones por paso. Cada una de esas operaciones es una suma, resta y multiplicación, que un procesador moderno ejecuta en nanosegundos.

Entonces, por cada paso de tiempo, el overhead de OpenMP es **mucho mayor** que el trabajo real que deben hacer los hilos. El resultado es que la versión paralela pasa la mayor parte del tiempo **gestionando hilos** en lugar de **calculando temperaturas**.

La paralelización solo es beneficiosa cuando el **trabajo por hilo es enorme** en comparación con el overhead. Es decir, cuando el problema es **masivo**.

Para el método de diferencias finitas, eso significa:

- `N_x` debe ser grande (miles o decenas de miles de puntos).
- `N_t` debe ser enorme (millones de pasos de tiempo).
- El producto `N_x * N_t` debe estar en el orden de **miles de millones** (10^9 o más).

Solo ahí el overhead se vuelve despreciable y se nota el speedup.

El código paralelo está **bien escrito**. La directiva `#pragma omp parallel for` es correcta y el algoritmo es paralelizable. El problema no es el código, es el **tamaño del problema**. Para problemas pequeños o medianos, el overhead de OpenMP domina y la versión serial es más rápida. La paralelización no es para todo; es una herramienta para problemas **realmente grandes** que tardan horas o días en ejecutarse en serial. Para este notebook, el serial es suficiente.

## **Conclusiones**

A lo largo de este notebook, hemos recorrido un camino que comenzó con una pregunta cotidiana —¿cuánto tarda el calor en propagarse por una barra?— y terminó con la implementación de un código paralelo en C usando OpenMP. En el medio, desarrollamos conceptos fundamentales de física, matemática y computación de alto rendimiento.

La ecuación del calor, propuesta por Fourier en el siglo XIX, captura de manera elegante un fenómeno universal: la temperatura en un punto cambia proporcionalmente a la curvatura del perfil térmico alrededor de ese punto. Esta relación, escrita como `∂u/∂t = α ∂²u/∂x²`, es una ecuación diferencial parcial que involucra derivadas tanto temporales como espaciales. Para resolverla numéricamente, recurrimos al método de diferencias finitas, que transforma el problema continuo en uno discreto: dividimos la barra en puntos y el tiempo en pasos, aproximando las derivadas mediante diferencias entre valores vecinos.

El código serial implementa el método explícito de diferencias finitas de forma directa. Con `N_x = 100` puntos y `N_t = 1000` pasos de tiempo, obtuvimos un error máximo de `2.78e-06`, una precisión más que suficiente para cualquier aplicación práctica. El coeficiente de estabilidad fue de `0.049`, muy por debajo del límite teórico de `0.5`, garantizando un comportamiento numérico estable.

La paralelización fue sorprendentemente sencilla: bastó con agregar la directiva `#pragma omp parallel for` antes del bucle que recorre los puntos espaciales. Esto se debe a que cada iteración es independiente: el cálculo de la temperatura en un punto solo requiere los valores de sus vecinos del paso anterior, sin dependencias entre puntos.

Sin embargo, los resultados mostraron que la versión paralela fue **significativamente más lenta** que la serial (62.93 ms vs 1.60 ms). Esto no indica un error en el código, sino un fenómeno conocido como **overhead**: la creación, distribución y sincronización de hilos tiene un costo fijo que, para problemas pequeños, supera cualquier ganancia. En nuestro caso, el trabajo útil por paso de tiempo era apenas de 100 operaciones, mientras que el overhead de OpenMP ocurre 1000 veces (un paso por cada instante de tiempo). El resultado es que el programa pasa más tiempo gestionando hilos que calculando temperaturas.

La paralelización es útil cuando el problema es **masivo**. Para que el speedup sea positivo, se necesita que el producto `N_x * N_t` esté en el orden de miles de millones de operaciones. Solo ahí el overhead se vuelve despreciable. En problemas pequeños o medianos, la versión serial es más eficiente.

Este enfoque nos enseñó que la optimización no es automática. No basta con agregar una directiva de paralelización; hay que entender el tamaño del problema, la naturaleza de las dependencias y los costos ocultos del hardware. La ecuación del calor es un ejemplo perfecto para aprender estas lecciones: es simple, didáctica y permite explorar conceptos profundos sin perderse en complejidades innecesarias.

El código desarrollado es funcional, preciso y está listo para ser adaptado a problemas más grandes. Y si algún día necesitamos simular una barra con una malla finísima y millones de pasos de tiempo, ya sabemos qué herramienta usar y cuándo.